Add SetFlag() function to set flags on/off via boolean.

binki opened the pull request:
Hi,

I was looking at making some code I was touching more readable/less error prone by taking a flags-enabled enum value, a particular flag, and a boolean to either set the bits for that flag on or off as a logical operation. I’m not sure if your library is attempting to solve that sort of problem, but I did find that building such a method on top of your idiom of dynamically compiling away `GetUnderlyingType()` calls provided a good starting point.

This adds a `BitwiseNot()` (for when turning the flag off / setting `on: false`) and `T SetFlag(T value, T flag, bool on = true)` the current way I envision this working.

If you consider incorporating this change, may you also consider publishing an assembly-oriented nuget package for this library? I’d like to use it and `SetFlag()` while letting nuget deal with the sort of thing it’s meant to deal with—referencing external libraries with reusable code that shouldn’t have to be recreated in every project.

binki committed:
Add SetFlag() function to set flags on/off via boolean.

arogozine commented on EnumUtil/EnumUtilBase.cs:
```diff
@@ -101,6 +105,9 @@ public abstract class EnumUtilBase<E>
         public static string[] GetNames<T>() where T : struct, E
             => Enum.GetNames(typeof(T));
 
+        public static T SetFlag<T>(T value, T flag, bool on = true) where T : struct, E
```
SetFlag should be split into two separate functions - Set and Unset or similar.

arogozine commented on EnumUtil/EnumCompiledCache.cs:
```diff
@@ -71,6 +71,26 @@ internal static class EnumCompiledCache<T>
                 .Compile();
         }
 
+        static Func<T, T> BitwiseUnaryOperator(ExpressionType expressionType)
```
Declare private modifier explicitly for consistency with the rest of the code.

binki commented:
> SetFlag should be split into two separate functions - Set and Unset or similar

What if I have a boolean variable and want to just make one call to set the flag on if the boolean is true and unset it if it is false, like how `jQuery.toggleClass()` with a second boolish argument works? I could add an UnsetFlag(x) which just calls SetFlag(x, false), would that be accepted?

> Declare private modifier explicitly for consistency with the rest of the code.

OK, let me do that…

binki committed:
Explicitly use private member decorator.

As requested by @arogozine.

binki committed:
Add UnsetFlag() as shorthand for SetFlag(orig, flag, false).

@arogozine wants SetFlag()/UnsetFlag() symmetric functions. I want the
ability to pass a boolean to a function (thus SetFlag() still has a
boolean on parameter indicating whether to set the flag ON or
OFF). Maybe this is an acceptable compromise?

binki commented:
I added in the explicit `private` modifiers, that is straightforward.

About the `SetFlag()`/`UnsetFlag()`, I added `UnsetFlag()` but did not remove the `on` parameter from `SetFlag()`. Thus you can use the method and pass in a bool indicating whether to set the flag ON or OFF _or_ you can ignore the extra parameter and use either `SetFlag()` or `UnsetFlag()`.

I would really like to keep around the passing in `bool` option. Otherwise I would have to write code like:

```
// This assumes that the bool parameter was removed from EnumUtil.SetFlag()
// so that both SetFlag() and UnsetFlag() share the same signature.
MyFlagProperty = (enable ? EnumUtil.SetFlag : EnumUtil.UnsetFlag)(MyFlagProperty, MyFlag.Flag);
```

where if I could just write something like:

```
MyFlagProperty = EnumUtil.SetFlag(MyFlagProperty, MyFlag.Flag, enable);
```

or perhaps

```
MyFlagProperty = EnumUtil.ToggleFlag(MyFlagProperty, MyFlag.Flag, enable);
```

Thoughts?

arogozine commented:
I see where you are coming from.

The function is a "Set" function with an parameter called "on". False value of "on" means "off".
Thus, you can use this function to "set a flag off" which is confusing to me :confused:

I think you are on to something.
`Toggle` sounds a lot better.

What about something like,

``` csharp
// Toggle on/off - using XOR
ToggleFlag<T>(T value, T flag) { /**/ }

// Toggle based on boolean parameter
ToggleFlag<T>(T value, T flag, bool flagSet)  { /**/ }
```

Also, I will research creating a NuGet package.

binki committed:
Move SetFlag() with bool functionality to new ToggleFlag().

Now SetFlag() and UnsetFlag() always only ever set or unset a
flag. There is a new ToggleFlag() method with two overloads. With a
boolean, it will call SetFlag() or UnsetFlag() appropriately for
you. Without a boolean it will unset the flag if it is set or set it
if it is unset. According to @arogozine, this is less confusing than
being able to set a flag to OFF with SetFlag().

binki commented:
OK, I’ve moved the boolean flag to a `ToggleFlag()` API as suggested.

binki committed:
More test assertions for ToggleFlag().

Now tests behavior when the initial value has more than one flag
set. I think this better verifies correct XOR behavior.

arogozine commented:
Looks good; merged.
